src/nix/search.cc

```diff
@@ -110,37 +145,47 @@ struct CmdSearch : InstallableCommand, MixJSON
                     std::replace(description.begin(), description.end(), '\n', ' ');
                     auto attrPath2 = concatStringsSep(".", attrPath);
 
-                    std::smatch attrPathMatch;
-                    std::smatch descriptionMatch;
-                    std::smatch nameMatch;
+                    std::vector<std::smatch> attrPathMatches;
+                    std::vector<std::smatch> descriptionMatches;
+                    std::vector<std::smatch> nameMatches;
+                    bool found = false;
 
                     for (auto & regex : regexes) {
-                        std::regex_search(attrPath2, attrPathMatch, regex);
-                        std::regex_search(name.name, nameMatch, regex);
-                        std::regex_search(description, descriptionMatch, regex);
-                        if (!attrPathMatch.empty()
-                            || !nameMatch.empty()
-                            || !descriptionMatch.empty())
-                            found++;
+                        found = false;
+                        auto add_all = [&found](std::sregex_iterator it, std::vector<std::smatch>& vec){
+                            const auto end = std::sregex_iterator();
+                            while(it != end) {
+                                vec.push_back(*it++);
+                                found = true;
+                            }
+                        };
```

Comment:
Rather than consuming the `sregex_iterator` and putting everything in a vector, isn’t it possible to keep the iterator and directly pass it to `hilite_all`? That would make the code both simpler and (I think) marginally more efficient

Reply:
That's a good idea but I think we would still need a vector of iterators and such a solution probably wouldn't be any simpler.

Reply:
Oh indeed you’re right. Would have loved to have some kind of combined iterator to transparently iterate over both, but since there’s nothing standard for that, let’s keep the vector